File: question/neetcode150/sliding_window/longest-substring-without-repeating-characters.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0;
        int n = s.size();
        unordered_set<char> d;
        int ans = 0;

        for (int r=0; r < n;r++){
            char c = s[r];
            while (d.count(c)){
                d.erase(s[l]);
                l++;
            }
            d.insert(c);
            ans = max(ans, r-l+1);
        }

        return ans;
    }
};

File: question/neetcode150/sliding_window/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        n = len(s)

        d = set()
        ans = 0
        for r in range(n):
            while s[r] in d:
                d.remove(s[l])
                l+=1
            d.add(s[r])
            ans = max(ans, r-l)
        
        return ans